File: app/src/main/java/com/radardeal/app/monitoring/KnownIdCache.kt
package com.radardeal.app.monitoring

import java.util.concurrent.ConcurrentHashMap

/**
 * The set of listing ids RadarDeal already knows, held in memory, per watch.
 *
 * This exists to keep Room off the detection path. The previous engine read the whole
 * `listings` table on **every** scan just to answer "have I seen this id before?" — at a
 * 3-second interval with a few hundred stored listings that is a database round trip and a
 * few hundred object allocations, several times a minute, for an answer that is almost always
 * "yes, all of them".
 *
 * Room remains the source of truth: the cache is warmed from it once per watch, and every
 * newly accepted id is written back asynchronously. The cache only ever *shortcuts* the
 * question, it never decides what is persisted.
 *
 * Thread safety: reads happen on whatever dispatcher a scan runs on and writes can arrive from
 * the WebView's live channel at the same time, so the backing sets are concurrent. A stale
 * "unknown" answer is harmless — the diff downstream is still authoritative and will simply
 * merge into the existing row instead of creating one.
 */
class KnownIdCache {

    private val byWatch = ConcurrentHashMap<Long, MutableSet<String>>()

    /** Watches whose set has been loaded from Room at least once. */
    private val warmed = ConcurrentHashMap.newKeySet<Long>()

    fun isWarm(watchId: Long): Boolean = warmed.contains(watchId)

    /** Replaces the set for [watchId] with [ids]. Called once, from the first scan. */
    fun warm(watchId: Long, ids: Collection<String>) {
        val set = ConcurrentHashMap.newKeySet<String>(maxOf(ids.size * 2, 64))
        set.addAll(ids)
        byWatch[watchId] = set
        warmed.add(watchId)
    }

    fun isKnown(watchId: Long, itemId: String): Boolean =
        byWatch[watchId]?.contains(itemId) == true

    /**
     * Returns the ids of [candidates] that this cache has never seen, preserving order.
     *
     * This is the first operation of every scan. When it returns an empty list — the common
     * case — the caller can stop immediately without building a single listing object.
     */
    fun filterUnknown(watchId: Long, candidates: List<String>): List<String> {
        val known = byWatch[watchId] ?: return candidates
        if (candidates.isEmpty()) return emptyList()
        var result: ArrayList<String>? = null
        for (id in candidates) {
            if (!known.contains(id)) {
                (result ?: ArrayList<String>(4).also { result = it }).add(id)
            }
        }
        return result ?: emptyList()
    }

    /**
     * Marks [itemId] as known and reports whether this call is the one that claimed it.
     *
     * Returning false means another thread — typically the live DOM channel and a polling scan
     * racing on the same listing — already took it, so the caller must not notify again.
     */
    fun claim(watchId: Long, itemId: String): Boolean {
        val set = byWatch.getOrPut(watchId) { ConcurrentHashMap.newKeySet() }
        return set.add(itemId)
    }

    fun addAll(watchId: Long, ids: Collection<String>) {
        if (ids.isEmpty()) return
        byWatch.getOrPut(watchId) { ConcurrentHashMap.newKeySet() }.addAll(ids)
    }

    fun size(watchId: Long): Int = byWatch[watchId]?.size ?: 0

    /** Called when a watch is deleted, or when its criteria change and the baseline resets. */
    fun forget(watchId: Long) {
        byWatch.remove(watchId)
        warmed.remove(watchId)
    }

    fun clear() {
        byWatch.clear()
        warmed.clear()
    }
}

File: ios/RadarDealCore/Sources/RadarDealCore/KnownIDCache.swift
import Foundation

/// The set of listing ids RadarDeal already knows, held in memory, per watch.
///
/// This keeps storage off the detection path. The question "have I seen this id before?" is
/// asked once per card on every scan; answering it from SwiftData would mean a fetch and a few
/// hundred object materialisations, several times a minute, for an answer that is almost always
/// "yes, all of them".
///
/// Storage remains the source of truth: the cache is warmed from it once per watch, and every
/// newly accepted id is written back afterwards. The cache only ever *shortcuts* the question.
///
/// Thread safety: a lock rather than an actor, because the hot path is synchronous and called
/// from a scan running off the main thread — making it an actor would force every lookup to
/// become a suspension point, which is precisely the cost this class exists to avoid.
public final class KnownIDCache: @unchecked Sendable {

    private let lock = NSLock()
    private var byWatch: [Int64: Set<String>] = [:]
    private var warmed: Set<Int64> = []

    public init() {}

    public func isWarm(_ watchID: Int64) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return warmed.contains(watchID)
    }

    /// Replaces the set for `watchID`. Called once, from the first scan.
    public func warm(_ watchID: Int64, ids: [String]) {
        lock.lock(); defer { lock.unlock() }
        byWatch[watchID] = Set(ids)
        warmed.insert(watchID)
    }

    public func isKnown(_ watchID: Int64, _ itemID: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return byWatch[watchID]?.contains(itemID) ?? false
    }

    /// Returns the ids of `candidates` this cache has never seen, preserving order.
    ///
    /// This is the first operation of every scan. When it returns an empty array — the common
    /// case — the caller can stop immediately without building a single listing.
    public func filterUnknown(_ watchID: Int64, _ candidates: [String]) -> [String] {
        lock.lock(); defer { lock.unlock() }
        guard let known = byWatch[watchID] else { return candidates }
        guard !candidates.isEmpty else { return [] }

        var result: [String] = []
        for id in candidates where !known.contains(id) {
            result.append(id)
        }
        return result
    }

    /// Marks `itemID` as known and reports whether this call is the one that claimed it.
    ///
    /// Returning false means another task already took it, so the caller must not notify again.
    @discardableResult
    public func claim(_ watchID: Int64, _ itemID: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return byWatch[watchID, default: []].insert(itemID).inserted
    }

    public func addAll(_ watchID: Int64, ids: [String]) {
        guard !ids.isEmpty else { return }
        lock.lock(); defer { lock.unlock() }
        byWatch[watchID, default: []].formUnion(ids)
    }

    public func size(_ watchID: Int64) -> Int {
        lock.lock(); defer { lock.unlock() }
        return byWatch[watchID]?.count ?? 0
    }

    /// Called when a watch is deleted, or when its criteria change and the baseline resets.
    public func forget(_ watchID: Int64) {
        lock.lock(); defer { lock.unlock() }
        byWatch.removeValue(forKey: watchID)
        warmed.remove(watchID)
    }

    public func clear() {
        lock.lock(); defer { lock.unlock() }
        byWatch.removeAll()
        warmed.removeAll()
    }
}
